File: c/real_self_energy.c
/* Copyright (C) 2015 Atsushi Togo */
/* All rights reserved. */

/* This file is part of phonopy. */

/* Redistribution and use in source and binary forms, with or without */
/* modification, are permitted provided that the following conditions */
/* are met: */

/* * Redistributions of source code must retain the above copyright */
/*   notice, this list of conditions and the following disclaimer. */

/* * Redistributions in binary form must reproduce the above copyright */
/*   notice, this list of conditions and the following disclaimer in */
/*   the documentation and/or other materials provided with the */
/*   distribution. */

/* * Neither the name of the phonopy project nor the names of its */
/*   contributors may be used to endorse or promote products derived */
/*   from this software without specific prior written permission. */

/* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS */
/* "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT */
/* LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS */
/* FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE */
/* COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, */
/* INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, */
/* BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; */
/* LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER */
/* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT */
/* LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN */
/* ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE */
/* POSSIBILITY OF SUCH DAMAGE. */

#include "real_self_energy.h"

#include <math.h>
#include <stdlib.h>

#include "phonoc_array.h"
#include "phonoc_utils.h"
#include "real_to_reciprocal.h"

static double get_real_self_energy_at_band(
    const long band_index, const Darray *fc3_normal_squared,
    const double fpoint, const double *frequencies, const long (*triplets)[3],
    const long *triplet_weights, const double epsilon, const double temperature,
    const double unit_conversion_factor, const double cutoff_frequency);
static double sum_real_self_energy_at_band(
    const long num_band, const double *fc3_normal_squared, const double fpoint,
    const double *freqs1, const double *freqs2, const double epsilon,
    const double temperature, const double cutoff_frequency);
static double sum_real_self_energy_at_band_0K(
    const long num_band, const double *fc3_normal_squared, const double fpoint,
    const double *freqs1, const double *freqs2, const double epsilon,
    const double cutoff_frequency);

void rse_get_real_self_energy_at_bands(
    double *real_self_energy, const Darray *fc3_normal_squared,
    const long *band_indices, const double *frequencies,
    const long (*triplets)[3], const long *triplet_weights,
    const double epsilon, const double temperature,
    const double unit_conversion_factor, const double cutoff_frequency) {
    long i, num_band0, num_band, gp0;
    double fpoint;

    num_band0 = fc3_normal_squared->dims[1];
    num_band = fc3_normal_squared->dims[2];
    gp0 = triplets[0][0];

    /* num_band0 and num_band_indices have to be same. */
    for (i = 0; i < num_band0; i++) {
        fpoint = frequencies[gp0 * num_band + band_indices[i]];
        if (fpoint < cutoff_frequency) {
            real_self_energy[i] = 0;
        } else {
            real_self_energy[i] = get_real_self_energy_at_band(
                i, fc3_normal_squared, fpoint, frequencies, triplets,
                triplet_weights, epsilon, temperature, unit_conversion_factor,
                cutoff_frequency);
        }
    }
}

void rse_get_real_self_energy_at_frequency_point(
    double *real_self_energy, const double frequency_point,
    const Darray *fc3_normal_squared, const long *band_indices,
    const double *frequencies, const long (*triplets)[3],
    const long *triplet_weights, const double epsilon, const double temperature,
    const double unit_conversion_factor, const double cutoff_frequency) {
    long i, num_band0;

    num_band0 = fc3_normal_squared->dims[1];

    /* num_band0 and num_band_indices have to be same. */
    for (i = 0; i < num_band0; i++) {
        if (frequency_point < cutoff_frequency) {
            real_self_energy[i] = 0;
        } else {
            real_self_energy[i] = get_real_self_energy_at_band(
                i, fc3_normal_squared, frequency_point, frequencies, triplets,
                triplet_weights, epsilon, temperature, unit_conversion_factor,
                cutoff_frequency);
        }
    }
}

static double get_real_self_energy_at_band(
    const long band_index, const Darray *fc3_normal_squared,
    const double fpoint, const double *frequencies, const long (*triplets)[3],
    const long *triplet_weights, const double epsilon, const double temperature,
    const double unit_conversion_factor, const double cutoff_frequency) {
    long i, num_triplets, num_band0, num_band, gp1, gp2;
    double shift;

    num_triplets = fc3_normal_squared->dims[0];
    num_band0 = fc3_normal_squared->dims[1];
    num_band = fc3_normal_squared->dims[2];

    shift = 0;
#ifdef _OPENMP
#pragma omp parallel for private(gp1, gp2) reduction(+ : shift)
#endif
    for (i = 0; i < num_triplets; i++) {
        gp1 = triplets[i][1];
        gp2 = triplets[i][2];
        if (temperature > 0) {
            shift += sum_real_self_energy_at_band(
                         num_band,
                         fc3_normal_squared->data +
                             i * num_band0 * num_band * num_band +
                             band_index * num_band * num_band,
                         fpoint, frequencies + gp1 * num_band,
                         frequencies + gp2 * num_band, epsilon, temperature,
                         cutoff_frequency) *
                     triplet_weights[i] * unit_conversion_factor;
        } else {
            shift +=
                sum_real_self_energy_at_band_0K(
                    num_band,
                    fc3_normal_squared->data +
                        i * num_band0 * num_band * num_band +
                        band_index * num_band * num_band,
                    fpoint, frequencies + gp1 * num_band,
                    frequencies + gp2 * num_band, epsilon, cutoff_frequency) *
                triplet_weights[i] * unit_conversion_factor;
        }
    }
    return shift;
}

static double sum_real_self_energy_at_band(
    const long num_band, const double *fc3_normal_squared, const double fpoint,
    const double *freqs1, const double *freqs2, const double epsilon,
    const double temperature, const double cutoff_frequency) {
    long i, j;
    double n1, n2, f1, f2, f3, f4, shift;
    /* double sum; */

    shift = 0;
    for (i = 0; i < num_band; i++) {
        if (freqs1[i] > cutoff_frequency) {
            n1 = phonoc_bose_einstein(freqs1[i], temperature);
            for (j = 0; j < num_band; j++) {
                if (freqs2[j] > cutoff_frequency) {
                    n2 = phonoc_bose_einstein(freqs2[j], temperature);
                    f1 = fpoint + freqs1[i] + freqs2[j];
                    f2 = fpoint - freqs1[i] - freqs2[j];
                    f3 = fpoint - freqs1[i] + freqs2[j];
                    f4 = fpoint + freqs1[i] - freqs2[j];

                    /* sum = 0;
                     * if (fabs(f1) > epsilon) {
                     *   sum -= (n1 + n2 + 1) / f1;
                     * }
                     * if (fabs(f2) > epsilon) {
                     *   sum += (n1 + n2 + 1) / f2;
                     * }
                     * if (fabs(f3) > epsilon) {
                     *   sum -= (n1 - n2) / f3;
                     * }
                     * if (fabs(f4) > epsilon) {
                     *   sum += (n1 - n2) / f4;
                     * }
                     * shift += sum * fc3_normal_squared[i * num_band + j]; */

                    shift +=
                        (-(n1 + n2 + 1) * f1 / (f1 * f1 + epsilon * epsilon) +
                         (n1 + n2 + 1) * f2 / (f2 * f2 + epsilon * epsilon) -
                         (n1 - n2) * f3 / (f3 * f3 + epsilon * epsilon) +
                         (n1 - n2) * f4 / (f4 * f4 + epsilon * epsilon)) *
                        fc3_normal_squared[i * num_band + j];
                }
            }
        }
    }
    return shift;
}

static double sum_real_self_energy_at_band_0K(
    const long num_band, const double *fc3_normal_squared, const double fpoint,
    const double *freqs1, const double *freqs2, const double epsilon,
    const double cutoff_frequency) {
    long i, j;
    double f1, f2, shift;

    shift = 0;
    for (i = 0; i < num_band; i++) {
        if (freqs1[i] > cutoff_frequency) {
            for (j = 0; j < num_band; j++) {
                if (freqs2[j] > cutoff_frequency) {
                    f1 = fpoint + freqs1[i] + freqs2[j];
                    f2 = fpoint - freqs1[i] - freqs2[j];
                    shift += (-1 * f1 / (f1 * f1 + epsilon * epsilon) +
                              1 * f2 / (f2 * f2 + epsilon * epsilon)) *
                             fc3_normal_squared[i * num_band + j];
                }
            }
        }
    }
    return shift;
}

File: rust/src/real_self_energy.rs
//! Real-part of the bubble-diagram self-energy.
//!
//! Port of `rse_get_real_self_energy_at_bands` and
//! `rse_get_real_self_energy_at_frequency_point` in
//! `c/real_self_energy.c`.  Frequencies and Bose-Einstein occupations
//! are in THz units throughout.
//!
//! `fc3_normal_squared` is the `|V^(3)|^2` array laid out as
//! ``(num_triplets, num_band0, num_band, num_band)`` in row-major
//! order.  Both entry points return `real_self_energy[0..num_band0]`
//! summed over triplets and scaled by `unit_conversion_factor`.

use rayon::prelude::*;

use crate::funcs::bose_einstein;

/// Finite-temperature inner sum at one triplet and one band0 index.
///
/// Mirrors `sum_real_self_energy_at_band` in `c/real_self_energy.c`.
/// The Cauchy principal value `1 / x` is regularised as
/// `x / (x^2 + epsilon^2)`.
fn sum_at_band(
    num_band: usize,
    fc3: &[f64],
    fpoint: f64,
    freqs1: &[f64],
    freqs2: &[f64],
    epsilon: f64,
    temperature_thz: f64,
    cutoff_frequency: f64,
) -> f64 {
    let eps2 = epsilon * epsilon;
    let mut shift = 0.0;
    for i in 0..num_band {
        let f1i = freqs1[i];
        if f1i <= cutoff_frequency {
            continue;
        }
        let n1 = bose_einstein(f1i, temperature_thz);
        let row = &fc3[i * num_band..(i + 1) * num_band];
        for j in 0..num_band {
            let f2j = freqs2[j];
            if f2j <= cutoff_frequency {
                continue;
            }
            let n2 = bose_einstein(f2j, temperature_thz);
            let f1 = fpoint + f1i + f2j;
            let f2 = fpoint - f1i - f2j;
            let f3 = fpoint - f1i + f2j;
            let f4 = fpoint + f1i - f2j;
            let d = -(n1 + n2 + 1.0) * f1 / (f1 * f1 + eps2)
                + (n1 + n2 + 1.0) * f2 / (f2 * f2 + eps2)
                - (n1 - n2) * f3 / (f3 * f3 + eps2)
                + (n1 - n2) * f4 / (f4 * f4 + eps2);
            shift += d * row[j];
        }
    }
    shift
}

/// Zero-temperature inner sum.  Mirrors
/// `sum_real_self_energy_at_band_0K` in `c/real_self_energy.c`.
fn sum_at_band_0k(
    num_band: usize,
    fc3: &[f64],
    fpoint: f64,
    freqs1: &[f64],
    freqs2: &[f64],
    epsilon: f64,
    cutoff_frequency: f64,
) -> f64 {
    let eps2 = epsilon * epsilon;
    let mut shift = 0.0;
    for i in 0..num_band {
        let f1i = freqs1[i];
        if f1i <= cutoff_frequency {
            continue;
        }
        let row = &fc3[i * num_band..(i + 1) * num_band];
        for j in 0..num_band {
            let f2j = freqs2[j];
            if f2j <= cutoff_frequency {
                continue;
            }
            let f1 = fpoint + f1i + f2j;
            let f2 = fpoint - f1i - f2j;
            let d = -f1 / (f1 * f1 + eps2) + f2 / (f2 * f2 + eps2);
            shift += d * row[j];
        }
    }
    shift
}

/// Per-band driver corresponding to `get_real_self_energy_at_band`.
/// Reduces across triplets in parallel.
#[allow(clippy::too_many_arguments)]
fn real_self_energy_at_band(
    band_index: usize,
    fc3_normal_squared: &[f64],
    num_triplets: usize,
    num_band0: usize,
    num_band: usize,
    fpoint: f64,
    frequencies: &[f64],
    triplets: &[[i64; 3]],
    triplet_weights: &[i64],
    epsilon: f64,
    temperature_thz: f64,
    unit_conversion_factor: f64,
    cutoff_frequency: f64,
) -> f64 {
    let stride_triplet = num_band0 * num_band * num_band;
    let stride_band0 = num_band * num_band;
    let band_base = band_index * stride_band0;

    (0..num_triplets)
        .into_par_iter()
        .map(|i| {
            let gp1 = triplets[i][1] as usize;
            let gp2 = triplets[i][2] as usize;
            let fc3 = &fc3_normal_squared
                [i * stride_triplet + band_base..i * stride_triplet + band_base + stride_band0];
            let freqs1 = &frequencies[gp1 * num_band..(gp1 + 1) * num_band];
            let freqs2 = &frequencies[gp2 * num_band..(gp2 + 1) * num_band];
            let inner = if temperature_thz > 0.0 {
                sum_at_band(
                    num_band,
                    fc3,
                    fpoint,
                    freqs1,
                    freqs2,
                    epsilon,
                    temperature_thz,
                    cutoff_frequency,
                )
            } else {
                sum_at_band_0k(
                    num_band,
                    fc3,
                    fpoint,
                    freqs1,
                    freqs2,
                    epsilon,
                    cutoff_frequency,
                )
            };
            inner * triplet_weights[i] as f64 * unit_conversion_factor
        })
        .sum()
}

/// Port of `rse_get_real_self_energy_at_bands`.  Writes
/// `real_self_energy[0..num_band0]` where band0 indices are taken from
/// `band_indices` and evaluated at the on-shell phonon frequency of the
/// first grid point of the triplet.
#[allow(clippy::too_many_arguments)]
pub(crate) fn real_self_energy_at_bands(
    real_self_energy: &mut [f64],
    fc3_normal_squared: &[f64],
    num_triplets: usize,
    num_band0: usize,
    num_band: usize,
    band_indices: &[i64],
    frequencies: &[f64],
    triplets: &[[i64; 3]],
    triplet_weights: &[i64],
    epsilon: f64,
    temperature_thz: f64,
    unit_conversion_factor: f64,
    cutoff_frequency: f64,
) {
    let gp0 = triplets[0][0] as usize;
    for i in 0..num_band0 {
        let bi = band_indices[i] as usize;
        let fpoint = frequencies[gp0 * num_band + bi];
        if fpoint < cutoff_frequency {
            real_self_energy[i] = 0.0;
        } else {
            real_self_energy[i] = real_self_energy_at_band(
                i,
                fc3_normal_squared,
                num_triplets,
                num_band0,
                num_band,
                fpoint,
                frequencies,
                triplets,
                triplet_weights,
                epsilon,
                temperature_thz,
                unit_conversion_factor,
                cutoff_frequency,
            );
        }
    }
}

/// Port of `rse_get_real_self_energy_at_frequency_point`.  Same as
/// `real_self_energy_at_bands` but all band0 entries share a single
/// external `frequency_point`.
#[allow(clippy::too_many_arguments)]
pub(crate) fn real_self_energy_at_frequency_point(
    real_self_energy: &mut [f64],
    frequency_point: f64,
    fc3_normal_squared: &[f64],
    num_triplets: usize,
    num_band0: usize,
    num_band: usize,
    frequencies: &[f64],
    triplets: &[[i64; 3]],
    triplet_weights: &[i64],
    epsilon: f64,
    temperature_thz: f64,
    unit_conversion_factor: f64,
    cutoff_frequency: f64,
) {
    for i in 0..num_band0 {
        if frequency_point < cutoff_frequency {
            real_self_energy[i] = 0.0;
        } else {
            real_self_energy[i] = real_self_energy_at_band(
                i,
                fc3_normal_squared,
                num_triplets,
                num_band0,
                num_band,
                frequency_point,
                frequencies,
                triplets,
                triplet_weights,
                epsilon,
                temperature_thz,
                unit_conversion_factor,
                cutoff_frequency,
            );
        }
    }
}
